store/tikv/latch/latch.go

```diff
@@ -0,0 +1,186 @@
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	timeout = startTs <= l.lastCommitTs
+	if timeout {
+		return
+	}
+	if len(l.waiting) == 0 || l.waiting[0] != startTs {
+		l.waiting = append(l.waiting, startTs)
+		newWait = true
+	}
+
+	acquire = l.waiting[0] == startTs
+	return
+}
+
+// release releases the transaction with startTs and commitTs from current latch.
+// isEmpty is true when the waiting queue is empty after release current transaction,
+// otherwise return the front transaction in queue.
+func (l *Latch) release(startTs uint64, commitTs uint64) (isEmpty bool, front uint64) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	if startTs != l.waiting[0] {
+		panic(fmt.Sprintf("invalid front ts %v, latch:%+v", startTs, l))
+	}
+	// for rollback or timeout, the commitTs maybe zero
+	if commitTs > l.lastCommitTs {
+		l.lastCommitTs = commitTs
+	}
+	l.waiting = l.waiting[1:]
+	if len(l.waiting) == 0 {
+		// Should we reset the lastCommitTs since next key
+		// may different from current key?
+		// l.lastCommitTs = 0
+		isEmpty = true
+	} else {
+		front = l.waiting[0]
+		isEmpty = false
+	}
+	return
+}
+
+// try append startTs into latch, return false
+// when startTs is already timeout
+func (l *Latch) append(startTs uint64) bool {
+	if startTs <= l.lastCommitTs {
+		return false
+	}
+	l.waiting = append(l.waiting, startTs)
+	return true
+}
+
+// Lock is the locks' information required for a transaction
+type Lock struct {
+	// the slot IDs of the latches(keys) that a startTs must acquire before being able to processed
+	requiredSlots []int
+	/// The number of latches that the transaction has acquired.
+	acquiredCount int
+	/// The number of latches that the waiting queue contains current transaction
+	waitedCount int
+	// Current transaction's startTs
+	startTs uint64
+}
+
+// NewLock creates a new lock
+func NewLock(startTs uint64, requiredSlots []int) Lock {
+	return Lock{
+		requiredSlots: requiredSlots,
+		acquiredCount: 0,
+		waitedCount:   0,
+		startTs:       startTs,
+	}
+}
+
+// Latches which are used for concurrency control
+// Each latch is indexed by a slit ID, hence the term latch and slot are used in interchangeable,
+// but conceptually a latch is a queue, and a slot is an index to the queue
+type Latches struct {
+	slots []Latch
+	size  int
+}
+
+// NewLatches the size will be rounded up to the power of 2
+func NewLatches(size int) Latches {
+	powerOfTwoSize := int(math.Pow(2, math.Ceil(math.Log2(float64(size)))))
+	slots := make([]Latch, powerOfTwoSize, powerOfTwoSize)
+	return Latches{
+		slots: slots,
+		size:  powerOfTwoSize,
+	}
+}
+
+// GenLock generates Lock for the transaction with startTs and keys
+func (latches *Latches) GenLock(startTs uint64, keys [][]byte) Lock {
+	hashes := make(map[int]bool)
+	for _, key := range keys {
+		hashes[latches.hash(key)] = true
+	}
+	slots := make([]int, 0, len(hashes))
+	for key := range hashes {
+		slots = append(slots, key)
+	}
+	sort.Ints(slots)
```

Comment:
why do we need hash value being sorted here?

Reply:
To avoid deadlock, for example:

* transactionA need (1,2,3)
* transactionB need (3,2,1)

And things go through the following steps:

1. Transaction A get (1,2)
2. Transaction B get (3)
3. Transaction A tries to get 3 failed, waiting for TransactionB 
4. Transaction B tries to get 2 failed, waiting for TransactionA